syntax = "proto3";

option java_package = "server";
option java_outer_classname = "ScravaProto";

service Scrava {
   rpc startConnection(EmptyMessage) returns (CipherUpdate) {}
   rpc login(ClientLogin) returns (ClientData) {}
   rpc register(ClientRegister) returns (ClientData) {}

   rpc saveProject(SudoMessage) returns (EmptyMessage) {}

   rpc getTutorialList(Query) returns (StringList) {}
   rpc getProjectList(Query) returns (StringList) {}

   rpc refreshCypher(EmptyMessage) returns (CipherUpdate) {}
   
   rpc getTutorial(Query) returns (SerializedObject) {}
   rpc getProject(Query) returns (SerializedObject) {}

}

message Query {
  string query = 1;
  int32 offset = 2;
}

message SudoMessage {
  string token = 1;
  string obj = 2;
}

message SerializedObject {
  int32 id = 1;
  string name = 2;
  string obj = 3;
}

message StringList {
  repeated string ids = 1;
  repeated string results = 2;
}

message CipherUpdate {
  bytes publicKey = 1;
}

message ClientData {
  int32 id = 1;
  string name = 2;
  string token = 3;
}

message ClientLogin {
  string name = 1;
  string password = 2;
}

message ClientRegister {
  string name = 1;
  string password = 2;
  string email = 3;
}

message StringResponse {
  string response = 1;
}

message EmptyMessage {}
